testing/sdk_tasks.py

```diff
@@ -57,16 +57,22 @@ def fn():
         prefix_clause = ''
         if prefix:
             prefix_clause = ' starting with "{}"'.format(prefix)
-        log.info('Waiting for {} tasks{} to have updated ids:\n- Old tasks: {}\n- Current tasks: {}'.format(
-            len(old_task_ids), prefix_clause, sorted(old_task_ids), sorted(task_ids)))
-
-        all_updated = True
-        for id in task_ids:
-            if id in old_task_ids:
-                all_updated = False
-        if len(task_ids) < len(old_task_ids):
-            all_updated = False
-        return all_updated
+
+        old_set = set(old_task_ids)
+        new_set = set(task_ids)
+        newly_launched_set = new_set.difference(old_set)
+        old_remaining_set = old_set.intersection(new_set)
+        all_updated = len(newly_launched_set) == len(new_set) and len(old_remaining_set) == 0
+        if all_updated:
+            return all_updated
+
+        # forgive the language a bit, but len('remained') == len('launched'),
+        # and similar for the rest of the label for task ids in the log line,
+        # so makes for easier reading
+        log.info('Waiting for tasks{} to have updated ids:\n- Old tasks (remained): {}\n- New tasks (launched): {}'.format(
```

Comment:
As it is now, we no longer log the success condition (since we return before this is the case).

Reply:
True. The success condition was previously logged as 'Waiting for ...', so prone to be removed as a "success". That said, I removed it intentionally.

We are not consistently logging to feed a state machine like observer that requires Enter/Exit <State> messages AND if we do, we should feed it better.

The logging here and throughout should be consistently providing either:
* Context that is useful if the system under test (or operation) were to crash or incorrectly operate.
* Context for a long-running process that may be under (typically human, most typically developer, but more importantly operator) observation.

Since this is test code logging, the operator is the developer so generally these reasons fold, yielding log in tests what provides feedback that the system isn't stalled or otherwise operating in a way that will lead to a sure fail.

Reply:
Fortunately, you were right @elezar . The method of expecting the resulting (new) task set to have an equal number of tasks to the previous (old) task set was actually being used (in one stray case, a check of placement constraints updated. Ideally, the tests would use a single method to ensure that the expected plan executed and completed, cycling the tasks in the process, but the tests are not at this point, so...re-instated the constraint.